### ISSUE TITLE
push.apply doesn't type check properly

### ISSUE THREAD

**Issue by yaru22** [google/closure-compiler#2016]
Hi,

In the following code:

``` js
function f() {
  var a = /** @type {Array<number>} */ ([1, 2, 3]);
  var b = /** @type {Array<string>} */ (['hello', 'world']);
  a.push.apply(a, b);  // doesn't give an error even though it should?
  a.push('hello');  // gives JSC_TYPE_MISMATCH as expected
}
```

I would expect `a.push.apply(a, b);` to give the type mismatch error but it doesn't. It looks like an error?

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250Afunction%2520f()%2520%257B%250A%2520%2520var%2520a%2520%253D%2520%252F**%2520%2540type%2520%257BArray%253Cnumber%253E%257D%2520*%252F%2520(%255B1%252C%25202%252C%25203%255D)%253B%250A%2520%2520var%2520b%2520%253D%2520%252F**%2520%2540type%2520%257BArray%253Cstring%253E%257D%2520*%252F%2520(%255B'hello'%252C%2520'world'%255D)%253B%250A%2520%2520a.push.apply(a%252C%2520b)%253B%2520%2520%252F%252F%2520doesn't%2520give%2520an%2520error%2520even%2520though%2520it%2520should%253F%250A%2520%2520a.push('hello')%253B%2520%2520%252F%252F%2520gives%2520JSC_TYPE_MISMATCH%2520as%2520expected%250A%257D%250A

**Comment by Dominator008** [comment]
Interesting, OTI gives a warning on the second push but not the first one, while NTI gives a warning on the first push but not the second one.
